proposals/file-types.md

```diff
@@ -0,0 +1,188 @@
+> - **Next, if the set of accessible members contains file types, all non-file types are removed from the set.**
+
+#### Remarks
+These rules disallow usage of file types outside the file in which they are declared.
+
+These rules also permit *shadowing* of a non-file type by a file type:
+```cs
+// File1.cs
+class C
+{
+    public static void M() { }
+}
+```
+
+```cs
+// File2.cs
+file class C
+{
+    public static void M() { }
+}
+
+class Program
+{
+    static void Main()
+    {
+        C.M(); // refers to the 'C' in File2.cs
+    }
+}
+```
+
+Note that we don't update the [scopes](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/basic-concepts.md#77-scopes) section of the spec. This is because, as the spec states:
+
+> The ***scope*** of a name is the region of program text within which it is possible to refer to the entity declared by the name without qualification of the name.
+
+In effect, scope only impacts the lookup of non-qualified names. This isn't quite the right concept for us to leverage because we need to also impact the lookup of qualified names:
+
+```cs
+// File1.cs
+namespace NS1
+{
+    file class C
+    {
+        public static void M() { }
+    }
+}
+
+namespace NS2
+{
+    class Program
+    {
+        public static void M()
+        {
+            C.M(); // error: C is not in scope
+            NS1.C.M(); // ok: C can be accessed through NS1.
+        }
+    }
+}
+```
+
+```cs
+// File2.cs
+namespace NS1
+{
+    class Program
+    {
+        C.M(); // error
+        NS1.C.M(); // error
+    }
+}
+```
+
+Therefore, we don't specify the feature in terms of which scope the type is contained in, but rather as additional "filtering rules" in member lookup.
+
+### Attributes
+In the initial iteration of the feature, a file type is not allowed to be used as an attribute. Such attributes are of limited utility for users who are unaware of the internal name generation strategy used in the implementation.
+
+```cs
+file class FC : System.Attribute { }
+
+[FC] // error
+class C
+{
+}
+```
+
+### Usage in signatures
+There is a general need to prevent `file` types from appearing in member signatures where the `file` type might not be in scope at the point of usage of the member.
+
+#### Only allow signature usage in members of `file` types
+Perhaps the simplest way to ensure this is to enforce that `file` types can only appear in signatures or as base types of other `file` types:
+
+```cs
+file class FileBase
+{
+}
+
+public class Derived : FileBase // error
+{
+    private FileBase M2() => new FileBase() // error
+}
+
+file class FileDerived : FileBase // ok
+{
+    private FileBase M2() => new FileBase() // ok
+}
+```
+
+Note that this does restrict usage in explicit implementations, even though such usages are safe. We do this in order to simplify the rules for the initial iteration of the feature.
+
+```cs
+file interface I
+{
+    void M(I i);
+}
+
+class C : I
```

Comment:
>: I [](http://example.com/codeflow?start=8&length=3)

Is it an error to implement the interface, even if there are no interface members?
```csharp
file interface I { }
public class C : I { } // error?
```

Reply:
It is not an error.